usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripFrequenciesOverlapTest.java

```diff
@@ -0,0 +1,285 @@
+/*
+ *  Copyright (c) 2020. MobilityData IO.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.mobilitydata.gtfsvalidator.usecase;
+
+import org.apache.logging.log4j.Logger;
+import org.junit.jupiter.api.Test;
+import org.mobilitydata.gtfsvalidator.domain.entity.gtfs.frequencies.Frequency;
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.base.Notice;
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.error.OverlappingTripFrequenciesNotice;
+import org.mobilitydata.gtfsvalidator.usecase.port.GtfsDataRepository;
+import org.mobilitydata.gtfsvalidator.usecase.port.ValidationResultRepository;
+import org.mobilitydata.gtfsvalidator.usecase.utils.TimeUtils;
+import org.mockito.ArgumentCaptor;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.*;
+
+class ValidateTripFrequenciesOverlapTest {
+
+    // suppressed warning regarding ignored result of method since it is not necessary here
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    @Test
+    void overlappingFrequenciesShouldGenerateNotice() {
+        // frequencies.txt
+        // | trip_id | start_time | end_time | headway_secs | exact_time |
+        // |---------|------------|----------|--------------|------------|
+        // | 12      | 12:00      | 18:00    | 600          | 0          |
+        // | 12      | 15:44      | 21:43    | 400          | 1          | this row overlaps with the previous one
+        // | 12      | 03:44      | 08:45    | 200          | 0          |
+
+        final GtfsDataRepository mockDataRepo = mock(GtfsDataRepository.class);
+        final ValidationResultRepository mockResultRepo = mock(ValidationResultRepository.class);
+        final Logger mockLogger = mock(Logger.class);
+        final TimeUtils mockTimeUtil = mock(TimeUtils.class);
+        when(mockTimeUtil.convertIntegerToHMMSS(1200)).thenReturn("1200");
+        when(mockTimeUtil.convertIntegerToHMMSS(1800)).thenReturn("1800");
+        when(mockTimeUtil.convertIntegerToHMMSS(1544)).thenReturn("1544");
+        when(mockTimeUtil.convertIntegerToHMMSS(2143)).thenReturn("2143");
+        when(mockTimeUtil.arePeriodsOverlapping(1200, 1800,
```

Comment:
I think mocking the implementation of the time utility class here is potentially dangerous, in that it is potentially giving the developer false confidence the permutations of the periods is passing when you don't in fact know if that is the case (because the actual implementation that determines this isn't a dependency in this module, but in the adapter module).

I'd prefer to see all of these tests be in the time utility implementation module, and limit the test for the use case to determine that each combination of frequencies are tested (e.g., the number of calls is correct).

Reply:
Following our discussion, I do see value in testing to make sure the method returns the proper notices based on the values the utility method returns. So I would say let's leave the code, but add explicit comments/Javadocs saying that the testing of the time check logic happens in the utility method test.